src/go/guestagent/pkg/iptables/iptables_test.go

```diff
@@ -0,0 +1,267 @@
+			},
+		},
+		{
+			name:       "With wildcard listener and valid port mappings",
+			listenerIP: net.IPv4(0, 0, 0, 0),
+			expectedEntries: []limaIptables.Entry{
+				{TCP: true, IP: net.IPv4(192, 168, 21, 10), Port: 1080},
+				{TCP: true, IP: net.IPv4(192, 168, 21, 11), Port: 1081},
+				{TCP: true, IP: net.IPv4(192, 168, 21, 12), Port: 1082},
+			},
+		},
+		{
+			name:       "With entries removed",
+			remove:     true,
+			listenerIP: net.IPv4(0, 0, 0, 0),
+			expectedEntries: []limaIptables.Entry{
+				{TCP: true, IP: net.IPv4(192, 168, 22, 10), Port: 1080},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 11), Port: 1081},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 12), Port: 1082},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 13), Port: 1083},
+				{TCP: true, IP: net.IPv4(192, 168, 22, 14), Port: 1084},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			iptablesScanner := fakeScanner{
+				expectedEntries: tt.expectedEntries,
+				expectedErr:     tt.expectedAddFuncErr,
+			}
+
+			testTracker := fakeTracker{
+				receivedID:          make(chan string),
+				receivedRemoveID:    make(chan string),
+				receivedPortMapping: make(chan nat.PortMap),
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			interval := time.Second
+			iptablesHandler := iptables.New(ctx, &testTracker, &iptablesScanner, tt.listenerIP, interval)
+
+			go func() {
+				require.NoError(t, iptablesHandler.ForwardPorts())
+				cancel()
+			}()
+
+			for i := 0; i < len(tt.expectedEntries); i++ {
+				id := <-testTracker.receivedID
+				expectedID := utils.GenerateID(entryToString(tt.expectedEntries[i]))
+				require.Equal(t, expectedID, id)
+
+				pm := <-testTracker.receivedPortMapping
+				portProto, err := nat.NewPort("tcp", strconv.Itoa(tt.expectedEntries[i].Port))
+				require.NoError(t, err)
+
+				expectedPortBinding := nat.PortBinding{
+					HostIP:   tt.listenerIP.String(),
+					HostPort: strconv.Itoa(tt.expectedEntries[i].Port),
+				}
+				require.Contains(t, pm[portProto], expectedPortBinding)
+			}
+
+			if tt.remove {
+				removedEntries := []limaIptables.Entry{
+					{TCP: true, IP: net.IPv4(192, 168, 22, 11), Port: 1081},
+					{TCP: true, IP: net.IPv4(192, 168, 22, 12), Port: 1082},
+				}
+				// update the entries
+				updatedEntries := []limaIptables.Entry{
+					{TCP: true, IP: net.IPv4(192, 168, 22, 10), Port: 1080},
+					{TCP: true, IP: net.IPv4(192, 168, 22, 13), Port: 1083},
+					{TCP: true, IP: net.IPv4(192, 168, 22, 14), Port: 1084},
+					{TCP: true, IP: net.IPv4(192, 168, 22, 15), Port: 1085},
+				}
+				iptablesScanner.expectedEntries = updatedEntries
+
+				for i := 0; i < len(removedEntries); i++ {
+					id := <-testTracker.receivedRemoveID
+					expectedID := utils.GenerateID(entryToString(removedEntries[i]))
+					require.Equal(t, expectedID, id)
+				}
+
+				addedElement := updatedEntries[len(updatedEntries)-1]
+				id := <-testTracker.receivedID
```

Comment:
The test should use some kind of timeout to cancel. Otherwise it could just hand forever, waiting for data on a channel that never arrives (when there is a bug in the code).